File: medium/RemoveNodeFromEndofLinkedList.cpp
#include <iostream>

using namespace std;
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *removeNthFromEnd(ListNode *head, int n) {

    ListNode dumy = ListNode(0);
    dumy.next = head;
    ListNode *f = &dumy, *m = head, *l = &dumy;
    int nth = n;

    while (nth > 0 and l->next != nullptr) {
      l = l->next;
      nth--;
    }
    while (l->next != nullptr) {
      f = f->next;
      m = m->next;
      l = l->next;
    }
    f->next = m->next;
    delete m;
    return dumy.next;
  }
};

File: py/medium/RemoveNodeFromEndofLinkedList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        left, right = dummy, head
        while n > 0:
            right = right.next  # type:ignore
            n -= 1
        while right:
            left = left.next  # type:ignore
            right = right.next  # type:ignore

        left.next = left.next.next  # type:ignore
        return dummy.next
